go/go116/Dockerfile

```diff
@@ -14,11 +14,27 @@
 
 FROM golang:1.16
 
-# Install pyenv
-RUN apt-get update
-RUN apt-get install -y --force-yes make build-essential libssl-dev zlib1g-dev libbz2-dev \
+# Install dependencies
+RUN set -ex; \
+  apt-get update -y; \
+  apt-get install -y \
+    make build-essential libssl-dev zlib1g-dev libbz2-dev \
     libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev libncursesw5-dev \
-    xz-utils tk-dev libffi-dev liblzma-dev python-openssl
+    xz-utils tk-dev libffi-dev liblzma-dev python-openssl \
+    apt-transport-https ca-certificates curl gnupg-agent lsb-release software-properties-common \
+    unzip wget vim; \
+    rm -rf /var/lib/apt/lists/*
+
+# Install docker
+RUN curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add - && \
```

Comment:
What is the use case for docker here?

Reply:
My test script runs `docker build` and other docker cli. 
I believe we did think about using cloud build but that increases runtime by 2x to 3x.

I do think think this docker in docker is available in [Java](https://github.com/googleapis/testing-infra-docker/blob/5ac625a7949d71846c1ea9306fe4267ae5abe0ba/java/java11/Dockerfile#L85), [Python](https://github.com/googleapis/testing-infra-docker/blob/5ac625a7949d71846c1ea9306fe4267ae5abe0ba/python/googleapis/python-multi/Dockerfile#L95), and I recently also added it to [node](https://github.com/googleapis/testing-infra-docker/blob/5ac625a7949d71846c1ea9306fe4267ae5abe0ba/node/14-user/Dockerfile#L32)

I may have to split up this PR since building 6 images times out.